chainer/optimizer.py

```diff
@@ -192,6 +192,10 @@ def update(self, param):
 
         self.t += 1
 
+        with chainer.using_device(param.device):
```

Comment:
@niboshi what do you think about https://github.com/chainer/chainer/pull/7590#discussion_r296987867 ?

Reply:
>How about setting the device where we're iterating over the parameters, i.e. inside the loop in `GradientMethod.update` ? We can unify this context and the `update_core`'s identical context?

It's the problem of responsibility. If we specify the caller of `UpdateRule.update`, and not the caller of `Optimizer.update`, to set the current device, your suggestion seems to make sense. But it might be asymmetric.

Reply:
According to our policy, `using_device` should be written in the lowest-level framework code that can call user-written function (in this case `UpdateRule.update_core`).
So current implementation is appropriate.